public/app/features/explore/LogsContainer.tsx

```diff
@@ -28,10 +27,78 @@ interface LogsContainerProps {
   onStopScanning: () => void;
 }
 
-export class LogsContainer extends PureComponent<PropsFromRedux & LogsContainerProps> {
-  onChangeTime = (absoluteRange: AbsoluteTimeRange) => {
+type LogsContainerState = {
+  logsToShow: LogsModel | null;
+  absoluteRangeToShow: AbsoluteTimeRange;
+};
+export class LogsContainer extends PureComponent<LogsContainerProps, LogsContainerState> {
+  private logRowsCache = new LRU<string, { cacheLogsResult: LogsModel; cacheAbsoluteRange: AbsoluteTimeRange }>(5);
+
+  constructor(props: LogsContainerProps) {
+    super(props);
+
+    this.state = {
+      logsToShow: props.logsResult,
+      absoluteRangeToShow: props.absoluteRange,
+    };
+  }
+
+  componentDidMount() {
+    const { logsResult, absoluteRange } = this.props;
+    if (logsResult && absoluteRange) {
+      this.setCacheLogResults(logsResult, absoluteRange);
+    }
+  }
+
+  componentDidUpdate(prevProps: LogsContainerProps) {
+    const { logsResult, absoluteRange } = this.props;
+    // If new results, update cache
+    if (logsResult && !isEqual(logsResult, prevProps.logsResult)) {
+      // If queries were changed, reset cache and start fresh
+      if (!isEqual(logsResult.queries, prevProps.logsResult?.queries)) {
+        this.logRowsCache.reset();
+      }
+      // Update state and add to logResults and absolutRange to chache
```

Comment:
I think it's spelled "cha-cha-cha" :D 

```suggestion
      // Update state and add to logResults and absolutRange to cache
```

![chachacha](https://user-images.githubusercontent.com/745532/117291160-eebe0c00-ae6e-11eb-9fcc-c9797cfcc1bf.gif)

Reply:
🤣🤣🤣🤣